src/cryptonote_core/blockchain.cpp

```diff
@@ -3312,118 +3313,201 @@ void Blockchain::flush_invalid_blocks() {
-                removable_reason = "node {}; height {} {}:{}"_format(
-                        protocol_dereg ? "dereg" : "expired",
-                        *unlock_or_dereg_height,
-                        epee::string_tools::get_ip_string_from_int32(ip),
-                        port);
-            } else {
-                removable_reason = "node exists only in contract";
-            }
+    // Incoming, unconfirmed registrations are not removable:
+    if (pending_reg) {
+        log::debug(
+                logcat,
+                "{} is not removable: found a pending but unconfirmed new service node event with "
+                "that BLS pubkey",
+                bls_pubkey);
+        return false;
+    }
 
-            fmt::format_to(
-                    std::back_inserter(buffer), " {:04d} {} ({})", index, it, removable_reason);
-        }
+    // Case 2: not in the registered SN list, recently removed, or incoming unconfirmed, but *is* in
+    // the contract: this is both removable and liquidatable.
+    if (service_node_list.find_public_key_registered(bls_pubkey)) {
+        log::debug(
+                logcat,
+                "{} is not removable: pubkey belongs to a registered service node",
+                bls_pubkey);
+        return false;
+    }
+
+    if (!m_l2_tracker)
+        log::debug(
+                logcat,
+                "No L2 tracker configured; skipping in-contract-but-not-oxend removable check of "
+                "{}",
+                bls_pubkey);
+
+    if (m_l2_tracker->is_in_contract(bls_pubkey).value_or(false)) {
+        log::debug(
+                logcat, "{} is removable: found in contract but not in oxend SN list", bls_pubkey);
+        return true;
     }
 
-    std::string log = fmt::to_string(buffer);
-    oxen::log::debug(logcat, "{}", fmt::to_string(buffer));
+    log::debug(
+            logcat,
+            "{} is not removable: BLS pubkey is neither registered, nor found in L2 contract",
+            bls_pubkey);
+    return false;
+}
+
+std::unordered_map<eth::bls_public_key, bool> Blockchain::get_removable_nodes() const {
+    std::unordered_map<eth::bls_public_key, bool> result;
+
+    const uint64_t height = get_current_blockchain_height();
+
+    log::debug(logcat, "Looking for removable nodes @ height {}", height);
+
+    // Case 1: recently removed nodes
+    size_t liq_count = 0;
+    service_node_list.for_each_recently_removed_node([&](const auto& node) {
+        assert(node.info.bls_public_key &&
+               "Invalid null key got inserted into the recently removed list");
+        const auto& blspk = node.info.bls_public_key;
+        bool liquidatable = height >= node.liquidation_height;
+        result.emplace(blspk, liquidatable);
+        if (liquidatable)
+            ++liq_count;
+        log::trace(
+                logcat,
+                "- {} @ {}:{} is {} (recently {} node @ height {})",
+                blspk,
+                node.public_ip,
```

Comment:
```suggestion
                epee::string_tools::get_ip_string_from_int32(node.public_ip),
```

Print as formatted IP address